Parsing string values in -D compiler option

Hi
I have compilation option -DPRODUCT_NAME='"SCTP IETF"' for my project. But when I’m using CodeChecker for code analysis, the following  error is raised:
sctp.c:715:1: error: use of undeclared identifier 'SCTP'
REGVERSION_DEF(SCTP)
version.h:66:36: note: expanded from macro 'REGVERSION_DEF'
                                   PRODUCT_NAME,                        \
                                   ^
<command line>:12:22: note: expanded from here
#define PRODUCT_NAME SCTP
Also I see that the following option is used in command line for clang
 -DPRODUCT_NAME="SCTP
Does CodeChecker support such parameter – string with space? 
I am using 5.6 version of CodeChecker
Best regards, Igor

--- a/libcodechecker/analyze/log_parser.py
+++ b/libcodechecker/analyze/log_parser.py
@@ -106,24 +106,20 @@ def parse_compile_commands_json(logfile, add_compiler_defaults=False):
             continue
 
         action = build_action.BuildAction(counter)
-
         if 'command' in entry:
             command = entry['command']
+
+            # Old versions of intercept-build (confirmed to those shipping
+            # with upstream clang-5.0) do escapes in another way:
+            # -DVARIABLE="a b" becomes -DVARIABLE=\"a b\" in the output.
+            # This would be messed up later on by options_parser, so need a
+            # fix here. (Should be removed once we are sure noone uses this
+            # intercept-build anymore!)
+            if r'\"' in command:
+                command = command.replace(r'\"', '"')
         elif 'arguments' in entry:
             # Newest versions of intercept-build create an argument vector
             # instead of a command string.
-            for i in range(0, len(entry['arguments'])):
-                arg = entry['arguments'][i]
-                if ' ' in arg:
-                    # If there is an argument with a space in it, the join
-                    # below will mess the invocation up. (-DVAR=va lue will be
-                    # passed as -DVAR=va)
-                    #
-                    # Build.json created by ld-logger escapes these strings
-                    # and they never reach the analyser later on.
-                    #
-                    # TODO: Better handle this. (See issue #505.)
-                    entry['arguments'][i] = '\"' + arg + '\"'
             command = ' '.join(entry['arguments'])
         else:
             raise KeyError("No valid 'command' or 'arguments' entry found!")
@@ -131,6 +127,7 @@ def parse_compile_commands_json(logfile, add_compiler_defaults=False):
 
         action.original_command = command
         action.analyzer_options = results.compile_opts
+
         action.lang = results.lang
         action.target = results.arch
 

--- a/libcodechecker/log/option_parser.py
+++ b/libcodechecker/log/option_parser.py
@@ -453,13 +453,21 @@ def parse_options(args):
     not only arguments."""
 
     # Keep " characters.
-    args = args.replace('"', '\\"')
+    args = args.replace(r'"', r'"\"')
 
     result_map = OptionParserResult()
 
     # The first element in the list is the compiler skip it from parsing.
     for it in OptionIterator(shlex.split(args)[1:]):
         arg_check(it, result_map)
+
+    for idx, opt in enumerate(result_map.compile_opts):
+        if '"' in opt:
+            # Arguments with a space in them are given back in a way
+            # (-DVAR="val ue") that they need an extra escape round. We can't
+            # do this earlier because shlex would undo these changes and
+            # mess up the result.
+            result_map.compile_opts[idx] = opt.replace('"', r'"\"')
 
     result_map.compiler = shlex.split(args)[0]
     is_source = False

--- a/tests/unit/logparser_test_files/intercept-new.json
+++ b/tests/unit/logparser_test_files/intercept-new.json
@@ -1,0 +1,14 @@
+[
+    {
+        "arguments": [
+            "c++",
+            "-c",
+            "-DVARIABLE=\"some value\"",
+            "-o",
+            "a.out",
+            "a.cpp"
+        ],
+        "directory": "/tmp",
+        "file": "a.cpp"
+    }
+]

--- a/tests/unit/logparser_test_files/intercept-old.json
+++ b/tests/unit/logparser_test_files/intercept-old.json
@@ -1,0 +1,7 @@
+[
+    {
+        "command": "c++ -c -DVARIABLE=\\\"some value\\\" -o a.out a.cpp",
+        "directory": "/tmp",
+        "file": "/tmp/a.cpp"
+    }
+]

--- a/tests/unit/logparser_test_files/ldlogger-new.json
+++ b/tests/unit/logparser_test_files/ldlogger-new.json
@@ -1,0 +1,7 @@
+[
+	{
+		"directory": "/tmp",
+		"command": "g++ -DVARIABLE=\"some value\" /tmp/a.cpp -o /tmp/a.out",
+		"file": "/tmp/a.cpp"
+	}
+]

--- a/tests/unit/logparser_test_files/ldlogger-old.json
+++ b/tests/unit/logparser_test_files/ldlogger-old.json
@@ -1,0 +1,7 @@
+[
+	{
+		"directory": "/tmp",
+		"command": "g++ \"-DVARIABLE=\"some value\"\" /tmp/a.cpp -o /tmp/a.out",
+		"file": "/tmp/a.cpp"
+	}
+]

--- a/tests/unit/test_log_parser.py
+++ b/tests/unit/test_log_parser.py
@@ -1,0 +1,112 @@
+# -----------------------------------------------------------------------------
+#                     The CodeChecker Infrastructure
+#   This file is distributed under the University of Illinois Open Source
+#   License. See LICENSE.TXT for details.
+# -----------------------------------------------------------------------------
+
+""" Test the log parser which builds build actions from JSON CCDBs. """
+
+import os
+import unittest
+
+from libcodechecker.analyze import log_parser
+from libcodechecker.log import option_parser
+
+
+class LogParserTest(unittest.TestCase):
+    """
+    Test the log parser which convers logfiles (JSON CCDBs) to build action
+    lists.
+    """
+
+    @classmethod
+    def setup_class(cls):
+        """Initialize test resources."""
+
+        # Already generated JSONs for the tests.
+        cls.__this_dir = os.path.dirname(__file__)
+        cls.__test_files = os.path.join(cls.__this_dir,
+                                        'logparser_test_files')
+
+    def test_old_ldlogger(self):
+        """
+        Test log file parsing escape behaviour with pre-2017 Q2 LD-LOGGER.
+        """
+        logfile = os.path.join(self.__test_files, "ldlogger-old.json")
+
+        # LD-LOGGER before http://github.com/Ericsson/codechecker/pull/631
+        # used an escape mechanism that, when parsed by the log parser via
+        # shlex, made CodeChecker parse arguments with multiword string
+        # literals in them be considered as "file" (instead of compile option),
+        # eventually ignored by the command builder, thus lessening analysis
+        # accuracy, as defines were lost.
+        #
+        # Logfile contains "-DVARIABLE="some value"".
+        #
+        # There is no good way to back-and-forth convert in log_parser or
+        # option_parser, so here we aim for a non-failing stalemate of the
+        # define being considered a file and ignored, for now.
+
+        build_action = log_parser.parse_log(logfile)[0]
+        results = option_parser.parse_options(build_action.original_command)
+
+        self.assertEqual(' '.join(results.files),
+                         r'"-DVARIABLE="some value"" /tmp/a.cpp')
+        self.assertEqual(len(build_action.analyzer_options), 0)
+
+    def test_new_ldlogger(self):
+        """
+        Test log file parsing escape behaviour with after-#631 LD-LOGGER.
+        """
+        logfile = os.path.join(self.__test_files, "ldlogger-new.json")
+
+        # LD-LOGGERS after http://github.com/Ericsson/codechecker/pull/631
+        # now properly log the multiword arguments. When these are parsed by
+        # the log_parser, the define's value will be passed to the analyzer.
+        #
+        # Logfile contains -DVARIABLE="some value".
+
+        build_action = log_parser.parse_log(logfile)[0]
+
+        self.assertEqual(list(build_action.sources)[0], r'/tmp/a.cpp')
+        self.assertEqual(len(build_action.analyzer_options), 1)
+        self.assertEqual(build_action.analyzer_options[0],
+                         r'-DVARIABLE="\"some value"\"')
+
+    def test_old_intercept_build(self):
+        """
+        Test log file parsing escape behaviour with clang-5.0 intercept-build.
+        """
+        logfile = os.path.join(self.__test_files, "intercept-old.json")
+
+        # Scan-build-py shipping with clang-5.0 makes a logfile that contains:
+        # -DVARIABLE=\"some value\"
+        #
+        # The define is passed to the analyzer properly.
+
+        build_action = log_parser.parse_log(logfile)[0]
+
+        self.assertEqual(list(build_action.sources)[0], r'/tmp/a.cpp')
+        self.assertEqual(len(build_action.analyzer_options), 1)
+        self.assertEqual(build_action.analyzer_options[0],
+                         r'-DVARIABLE="\"some value"\"')
+
+    def test_new_intercept_build(self):
+        """
+        Test log file parsing escapes with upstream (GitHub) intercept-build.
+        """
+        logfile = os.path.join(self.__test_files, "intercept-new.json")
+
+        # Upstream scan-build-py creates an argument vector, as opposed to a
+        # command string. This argument vector contains the define as it's
+        # element in the following format:
+        # -DVARIABLE=\"some value\"
+        #
+        # The define is passed to the analyzer properly.
+
+        build_action = log_parser.parse_log(logfile)[0]
+
+        self.assertEqual(list(build_action.sources)[0], r'/tmp/a.cpp')
+        self.assertEqual(len(build_action.analyzer_options), 1)
+        self.assertEqual(build_action.analyzer_options[0],
+                         r'-DVARIABLE="\"some value"\"')

